### PR TITLE
Add support for all Auth login methods in useAuth0 hook

### PR DESCRIPTION
**Pull request by sqiou11**
### Changes

Add functions to the hook that initiate and complete the various login processes supported by Auth. Specifically:
- `sendSMSCode` - uses `Auth.passwordlessWithSMS` to send a code.
- `authorizeWithSMS` - uses `Auth.loginWithSMS` to login using previously sent code, other than different input parameters its intended to function just like `authorize`.
- `sendEmailCode` - uses `Auth.passwordlessWithEmail` to send a code.
- `authorizeWithEmail` - uses `Auth.loginWithEmail` to login with with code.
- `sendMultifactorChallenge` - uses `Auth.multifactorChallenge` to send a code.
- `authorizeWithOOB` - uses `Auth.loginWithOOB` to login with code.
- `authorizeWithOTP` - uses `Auth.loginWithOTP` to login with code.
- `authorizeWithRecoveryCode` - uses `Auth.loginWithRecoveryCode` to login with code.

This change is inspired by my own use-case in a project, where I originally duplicated some of the source code here and modified it locally to login using passwordless SMS, but I figured it could be useful to others as well. I'm quite new to React so happy to hear any feedback!

### Testing

Unit tests are added to test these functions in use-auth0.spec.js, modeled after existing unit tests for `authorize`.

- [x] This change adds unit test coverage
- [x] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] All existing and new tests complete without errors
- [x] All active GitHub checks have passed

### CODE REVIEW

**Comment by poovamraj** [comment]
@sqiou11 this is a great PR indeed, but adding such methods to hooks would mean we would expose all the different kinds of authorization which is a lot and might bloat the API. that is why we were thinking to keep it to a minimum with the hooks and use the `auth0` object if more customization is required.

However I would like to keep this PR open to see if the community finds it as a requirement. 

Thank you for contributing to our repo :)

**Comment by tonydaly** [comment]
We're trying to do the same flow and this PR would definitely be useful.

@poovamraj if this PR isn't merged what's the best way to achieve the same result with the current hooks?

I've setup my `Auth0Provider`, but once I've done `auth0.auth.loginWithSMS` and `auth0.credentialsManager.saveCredentials` the hooks don't automatically update? However, once I do a refresh then everything works.

I've had to create a new Auth0 client to call these methods as they don't seem to be exposed by `useAuth0`.

Any help appreciated!

**Comment by sqiou11** [comment]
Thanks @poovamraj for the reply! Totally understand the API concerns. Would a separate hook+context+provider dedicated to passwordless authentication be better? That's essentially what I've done in my project so if that's the intention in this library and what you meant when you said "use the `auth0` object if more customization is required" then good to know :).

@tonydaly not sure what you mean by "setup my `Auth0Provider` but I don't think using the existing provider and then additionally creating a new Auth0 client to use `loginWithSMS` will work. The existing `Auth0Provider` creates it's own Auth0 client underneath, plus your new Auth0 client isn't tied to the existing `Auth0Provider`'s state. What I did was define my own version of everything - `Auth0Provider`, `Auth0Context`, and `useAuth0`.

**Comment by poovamraj** [comment]
@sqiou11 A seperate hook for all the Authentication APIs is definitely one we are considering, but we are still worried about the following concerns
- Would we have common state like `user` `error` `isInitilaized` in this hook as well?
- If we want the user to use these from the older `useAuth0` hook, then will it be confusing to call a method from one hook and expect the error in other hook
- Is it better to avoid all these confusions and just add all the authentication APIs to the current hook itself.

I'd love to hear everyone's thoughts on this as well.

**Comment by poovamraj** [comment]
@tonydaly as @sqiou11 suggested at this point the library doesn't support hooks with Authentication APIs. You have to setup the flow yourself.

We came to this conclusion considering projects which does embedded login would be complex enough that they would like to manage their own states. But looks like this would be a good addition to our Hooks API. We would like to consider how we want to design this API before starting on it!

**Comment by poovamraj** [comment]
Hi @sqiou11, we would like to introduce all our APIs through the single hook. Will it be possible for you to update the PR with all the other authentication endpoints as well?

**Comment by sqiou11** [comment]
> 

That's great to hear! I will work on updating the PR accordingly.

**Comment by poovamraj** [comment]
Thanks a lot @sqiou11 thats great of you 🙏

**Comment by poovamraj** [comment]
Hi @sqiou11 I see that you have done some amazing work in the PR. You think it is ready for review now?

**Comment by sqiou11** [comment]
> Hi @sqiou11 I see that you have done some amazing work in the PR. You think it is ready for review now?

Yep! It’s ready for review

**Review by poovamraj — CHANGES_REQUESTED** [review]
This is great work! I have added few NITs to the review. 

The name change to the methods are great. It is better than the previous naming we have used. Since we won't be able to change the naming once we release this, I wanted to check with the team internally as well to have more eyes on this before we go ahead. Will keep you in the loop about this though.

Thanks again for this amazing PR!

File: src/hooks/auth0-provider.js [hunk 1]
```diff
@@ -121,6 +126,142 @@ const Auth0Provider = ({domain, clientId, children}) => {
...
+      try {
+        const params = options.length ? options[0] : {};
+
+        const credentials = await client.auth.loginWithOOB(params);
+        const user = getIdTokenProfileClaims(credentials.idToken);
+
+        await client.credentialsManager.saveCredentials(credentials);
+        dispatch({type: 'LOGIN_COMPLETE', user});
+      } catch (error) {
+        console.log(error);
```
[the remark this answers is not in the document]
Comment (sqiou11) [reply src/hooks/auth0-provider.js]: Removed

File: src/hooks/auth0-provider.js [hunk 2]
```diff
@@ -121,6 +126,142 @@ const Auth0Provider = ({domain, clientId, children}) => {
...
+        const params = options.length ? options[0] : {};
+        params.scope = finalizeScopeParam(params?.scope);
+
+        const credentials = await client.auth.loginWithEmail(params);
+        const user = getIdTokenProfileClaims(credentials.idToken);
+
+        await client.credentialsManager.saveCredentials(credentials);
+        dispatch({type: 'LOGIN_COMPLETE', user});
+      } catch (error) {
+        console.log(error);
```
[the remark this answers is not in the document]
Comment (sqiou11) [reply src/hooks/auth0-provider.js]: Should be cleaned up now

**Comment by sqiou11** [comment]
Sounds good!

**Comment by evansims** [comment]
(Just giving the CI a kick to resolve some stalled tests.)

**Comment by andrewflo** [comment]
Would love to see this change :) Great work and looking forward to seeing if this one gets merged!

**Comment by poovamraj** [comment]
@SafiUllahAshfaq This is a valid feedback. We will explore supporting this and keep you updated

@andrewflo We are actively looking into this feature and we should be merging it as part of our next major version of this SDK.

**Comment by sqiou11** [comment]
@poovamraj since it's my first time creating a PR for a repo that isn't mine, just want to make sure that I'm doing things correctly here. I resolved your comments earlier with the extraneous log statements. Is there something else on my end that you're waiting on?

**Comment by poovamraj** [comment]
Hi @sqiou11, this is a great contribution to our project and we are really sorry this is taking a long time to get merged. The plan is to merge it along with our v3 release. We have plans to pick this PR up again in the coming weeks. Again, sorry for the delay. This is some great work here. Thank you 🙏

**Comment by sqiou11** [comment]
> Hi @sqiou11, this is a great contribution to our project and we are really sorry this is taking a long time to get merged. The plan is to merge it along with our v3 release. We have plans to pick this PR up again in the coming weeks. Again, sorry for the delay. This is some great work here. Thank you 🙏

No problem! Take all the time you need, just wanted to make sure I wasn't blocking you all.

**Review by frederikprijck — DISMISSED** [review]

**Comment by poovamraj** [comment]
Thanks a lot @sqiou11! Without you this wouldn't have been possible 👍 🙇 We really appreciate your contribution to our repo!

**Pull request merged**